test/snippet/read_mapper/mapper_final.cpp

```diff
@@ -0,0 +1,150 @@
+    std::filesystem::path index_path{};
+    std::filesystem::path sam_path{"out.sam"};
+    uint8_t errors{0};
+};
+
+struct reference_store_t
+{
+    std::vector<std::string> ids;
+    std::vector<std::vector<seqan3::dna4>> sequences;
+};
+
+struct dna4_traits : seqan3::sequence_file_input_default_traits_dna
+{
+    using sequence_alphabet = seqan3::dna4;
+};
+
+void read_reference(std::filesystem::path const & reference_path, reference_store_t & storage)
+{
+    for (auto && record : seqan3::sequence_file_input<dna4_traits>{reference_path})
+    {
+        storage.ids.push_back(std::move(record.id()));
+        storage.sequences.push_back(std::move(record.sequence()));
+    }
+}
+
+void map_reads(cmd_arguments const & arguments,
+               reference_store_t const & storage)
+{
+    // we need the alphabet and text layout before loading
+    seqan3::bi_fm_index<seqan3::dna4, seqan3::text_layout::collection> index;
+    {
+        std::ifstream is{arguments.index_path, std::ios::binary};
+        cereal::BinaryInputArchive iarchive{is};
+        iarchive(index);
+    }
+
+    seqan3::sequence_file_input<dna4_traits> query_file_in{arguments.query_path};
+
+    using sam_fields_t = seqan3::fields<seqan3::field::seq,
+                                        seqan3::field::id,
+                                        seqan3::field::ref_id,
+                                        seqan3::field::ref_offset,
+                                        seqan3::field::alignment,
+                                        seqan3::field::qual,
+                                        seqan3::field::mapq>;
+
+    seqan3::sam_file_output sam_out{arguments.sam_path, sam_fields_t{}};
+
+    namespace sc = seqan3::search_cfg;
+    seqan3::configuration const search_config = sc::max_error_total{sc::error_count{arguments.errors}}
+                                              | sc::hit_all_best{};
+
+    namespace ac = seqan3::align_cfg;
+    seqan3::configuration const align_config = ac::method_global{ac::free_end_gaps_sequence1_leading{true},
+                                                                 ac::free_end_gaps_sequence2_leading{false},
+                                                                 ac::free_end_gaps_sequence1_trailing{true},
+                                                                 ac::free_end_gaps_sequence2_trailing{false}}
+                                             | ac::edit_scheme
+                                             | ac::output_alignment{}
+                                             | ac::output_begin_position{}
+                                             | ac::output_score{};
+
+    for (auto && record : query_file_in)
+    {
+        auto const & query = record.sequence();
+        for (auto && result : search(query, index, search_config))
+        {
+            size_t const start = result.reference_begin_position() ? result.reference_begin_position() - 1 : 0;
```

Comment:
TODO: I think this should be `arguments.errors` instead of `1`?

Reply:
but then we need to check whether `begin_position` <= `arguments.errors`, right?

to detect underflows?

Reply:
Yup.

But I think I need to double-check why we need to this at all...

Reply:
I think we do not need to shift the reference view to the left (-1).